File: Exercise_1.py
# Time complexity: Amortized time complexity O(1) and Worst case O(n) for push and pop.
# Space complexity: O(1)
# Does it work on Leetcode: Yes
# Did you face any problems while doing this: No


# We use a push stack and a pop stack. 
# We only pop from the pop stack and push to the push stack
# If pop stack is empty, we move all elements from push stack into the pop stack for all future pops (until empty again)

class MyQueue:

    def __init__(self):
        self.pushstack = []
        self.popstack = []

    def push(self, x: int) -> None:
        self.pushstack.append(x)

    def pop(self) -> int:
        if not self.popstack:
            while self.pushstack:
                self.popstack.append(self.pushstack.pop())
        return self.popstack.pop()

    def peek(self) -> int:
        if self.popstack:
            return self.popstack[len(self.popstack)-1]
        if self.pushstack:
            return self.pushstack[0]
        

    def empty(self) -> bool:
        return not (self.popstack or self.pushstack)
        

File: Exercise_1.cpp
// Implement Queue using Stacks
// Time Complexity : O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Approach
// Create 2 stacks; in stack & out stack. All push happens in in stack.
// When a pop happens, transfer all elements from in stack to out stack and then pop the element.
// In some cases, when the out stack is empty and we need to do pop, then again push all the elements from in to out stack if in stack is not empty.

class MyQueue {
private: 
    stack<int> in;
    stack<int> out;
public:
    MyQueue() {
        
    }
    
    void push(int x) {
        in.push(x);
    }
    
    int pop() {
        peek();
        int popped = out.top();
        out.pop();
        return popped;
    }
    
    int peek() {
        if(out.empty()) {
            while(!in.empty()) {
                out.push(in.top());
                in.pop();
            }
        }
        return out.top();
    }
    
    bool empty() {
        return in.empty() && out.empty();
    }
};
